lib/ClangImporter/ClangImporter.cpp

```diff
@@ -4066,3 +4065,30 @@ swift::getModuleCachePathFromClang(const clang::CompilerInstance &Clang) {
   return llvm::sys::path::parent_path(SpecificModuleCachePath).str();
 }
 
+clang::FunctionDecl *ClangImporter::generateCXXFunctionTemplateSpecializations(
+    ASTContext &ctx, clang::FunctionTemplateDecl *func, SubstitutionMap subst) {
+  SmallVector<clang::TemplateArgument, 4> templateSubst;
+  std::unique_ptr<TemplateInstantiationError> error =
+      ctx.getClangTemplateArguments(func->getTemplateParameters(),
+                                    subst.getReplacementTypes(), templateSubst);
+  if (error) {
+    std::string failedTypesStr;
+    llvm::raw_string_ostream failedTypesStrStream(failedTypesStr);
+    llvm::interleaveComma(error->failedTypes, failedTypesStrStream);
+    // TODO: Use the location of the apply here.
+    // TODO: This error message should not reference implementation details.
+    // See: https://github.com/apple/swift/pull/33053#discussion_r477003350
+    ctx.Diags.diagnose(SourceLoc(),
+                       diag::unable_to_convert_generic_swift_types.ID,
+                       {func->getName(), StringRef(failedTypesStr)});
```

Comment:
@varungandhi-apple can I get your opinion on error handling again here? The issue is, on Windows, clang will crash if we don't exit here (because of an assertion: `TPL->size() == TemplateArgs.size() && "size mismatch between args and parms!"`). That's only on Windows, though. On other platforms, we're allowed to continue which I think is good because it allows us to report more errors before aborting compilation. I see a few paths forward, what do you think would be the best (others, please feel free to chime in as well)?
 1. Create an SR/TODO and don't test this particular error on Windows.
 2. Try to "fix" this in LLVM. I'm not sure this is a very good or practical solution, I think clang/llvm reviewers will probably be against removing this (completely valid) assertion for good reason.
 3. (somehow) Tell the compiler to abort compilation after emitting this error.

Reply:
[I only briefly skimmed the Clang code...] The assertion you pointed out is in `MicrosoftCXXNameMangler::mangleTemplateArgs`. I don't understand why the name mangler is being called upon template instantiation. That seems strange... shouldn't name mangling happen much later in the pipeline?

Reply:
Thanks, your comment made me realize, I think the issue is completely my fault. What we need to do is return `nullptr` after emitting the diagnostic. I think that will fix this. 

I think name mangling probably happens when we try to instantiate the function template.

Reply:
Ah. That's not what I was talking about (and I totally missed that we didn't return here in case of an error)... but glad that it helped you figured out the issue. 😅

Reply:
All is well that ends well :) Now let's just hope the build succeeds on Windows.